"panic: interface conversion: interface {} is []interface {}, not string" in replaceVars, at refvartransformer.go:43

### Summary
The variable replacement routine panics when given invalid content. Having an error is OK but not telling the user where the input is invalid isn't OK. 

### kustomize version
{Version:3.5.4 GitCommit:3af514fa9f85430f0c1557c4a0291e62112ab026 BuildDate:2020-01-17T14:28:58+00:00 GoOs:darwin GoArch:amd64}

### What happens
kustomize fails with panic. No output generated. No info as to cause of panic.

### What should happen
panic should be rescued and error returned indicating the problematic input text. For the example below this might be `Searching for variable replacements,  expected array of strings, found [["mistakenly use a array value"]].`

### Steps to Reproduce
Starting with valid yaml so it will pass the checker:
```
cat <<EOF > pod.yaml
apiVersion: core/v1
kind: Pod
metadata:
  name: kbug
  namespace: default
spec:
  containers:
    - name: kbug
      command:
        - replaceme
EOF
```
kustomization has a patch and also needs to have variable in it in order to execute the replaceVars function:
```
cat <<EOF > kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
configMapGenerator:
  - name: global
    literals:
      - blah=foobar
vars:
  - name: FOO_BAR
    objref:
      apiVersion: v1
      kind: ConfigMap
      name: global
    fieldref:
      fieldpath: data.blah
resources:
  - pod.yaml
patchesJson6902:
  - target:
      group: core
      version: v1
      kind: Pod
      name: kbug
      namespace: default
    path: badreplace.yaml
EOF
```
Patch file uses an incorrect replacement value by mistake
```
cat <<EOF > badreplace.yaml
- op: replace
  path: /spec/containers/0/command/0
  value:
     - mistakenly use a array value
EOF
```
The correct syntax would be 
```
- op: replace
  path: /spec/containers/0/command/0
  value:
     correctly use a string value
```
Then running the build results in a panic
```
kustomize build .
panic: interface conversion: interface {} is []interface {}, not string

goroutine 1 [running]:
sigs.k8s.io/kustomize/api/internal/accumulator.(*refVarTransformer).replaceVars(0xc0005eddd0, 0x50dd560, 0xc0005f5880, 0xc0001990e0, 0x7, 0xc0001990e8, 0x4513717)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/accumulator/refvartransformer.go:43 +0x639
sigs.k8s.io/kustomize/api/transform.MutateField(0xc0005edb30, 0xc0005ede80, 0x1, 0x1, 0xc000199200, 0xc0005759b8, 0x1, 0x1, 0xc0005ede60, 0x30)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/transform/mutatefield.go:36 +0x185
sigs.k8s.io/kustomize/api/transform.MutateField(0xc0005edb00, 0xc0005ede70, 0x2, 0x2, 0xc000198f00, 0xc0005759b8, 0x1, 0x1, 0x2, 0x3)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/transform/mutatefield.go:61 +0x45a
sigs.k8s.io/kustomize/api/transform.MutateField(0xc0005eda70, 0xc0005ede60, 0x3, 0x3, 0xc0002e6d00, 0xc0005759b8, 0x1, 0x1, 0x0, 0xc0005ede60)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/transform/mutatefield.go:53 +0x622
sigs.k8s.io/kustomize/api/internal/accumulator.(*refVarTransformer).Transform(0xc0005eddd0, 0x55fc180, 0xc00013ad20, 0x0, 0xc000575b70)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/accumulator/refvartransformer.go:104 +0x3bc
sigs.k8s.io/kustomize/api/internal/accumulator.(*ResAccumulator).Transform(...)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/accumulator/resaccumulator.go:138
sigs.k8s.io/kustomize/api/internal/accumulator.(*ResAccumulator).ResolveVars(0xc00013ad00, 0x0, 0x0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/accumulator/resaccumulator.go:151 +0x126
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).makeCustomizedResMap(0xc000255a40, 0x1, 0x0, 0x55ef820, 0x6278b80, 0x55ca800)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/target/kusttarget.go:140 +0x9c
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).MakeCustomizedResMap(...)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/internal/target/kusttarget.go:110
sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).Run(0xc000575cf8, 0x7ffeefbfee58, 0x1, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/sigs.k8s.io/kustomize/api@v0.3.2/krusty/kustomizer.go:77 +0x3d8
sigs.k8s.io/kustomize/kustomize/v3/internal/commands/build.(*Options).RunBuild(0xc000307830, 0x5586880, 0xc0000d0008, 0x1, 0x0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/kustomize/internal/commands/build/build.go:124 +0x86
sigs.k8s.io/kustomize/kustomize/v3/internal/commands/build.NewCmdBuild.func1(0xc0003a7b80, 0xc000269800, 0x1, 0x1, 0x0, 0x0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/kustomize/internal/commands/build/build.go:68 +0xab
github.com/spf13/cobra.(*Command).execute(0xc0003a7b80, 0xc0002697d0, 0x1, 0x1, 0xc0003a7b80, 0xc0002697d0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003a7680, 0x9, 0x55d2100, 0xc0003a81a0)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/kustomize-20200117-60956-18qh1a4/.brew_home/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	/private/tmp/kustomize-20200117-60956-18qh1a4/kustomize/main.go:23 +0x71
```

--- a/api/internal/accumulator/refvartransformer.go
+++ b/api/internal/accumulator/refvartransformer.go
@@ -40,7 +40,11 @@ func (rv *refVarTransformer) replaceVars(in interface{}) (interface{}, error) {
 	case []interface{}:
 		var xs []interface{}
 		for _, a := range in.([]interface{}) {
-			xs = append(xs, expansion2.Expand(a.(string), rv.mappingFunc))
+			x, ok := a.(string)
+			if !ok {
+				return nil, fmt.Errorf("expected array of strings, found %v", in)
+			}
+			xs = append(xs, expansion2.Expand(x, rv.mappingFunc))
 		}
 		return xs, nil
 	case map[string]interface{}:
@@ -49,7 +53,7 @@ func (rv *refVarTransformer) replaceVars(in interface{}) (interface{}, error) {
 		for k, v := range inMap {
 			s, ok := v.(string)
 			if !ok {
-				// This field not contain a $(VAR) since it is not
+				// This field can not contain a $(VAR) since it is not
 				// of string type. For instance .spec.replicas: 3 in
 				// a Deployment object
 				xs[k] = v
@@ -64,7 +68,7 @@ func (rv *refVarTransformer) replaceVars(in interface{}) (interface{}, error) {
 	case interface{}:
 		s, ok := in.(string)
 		if !ok {
-			// This field not contain a $(VAR) since it is not of string type.
+			// This field can not contain a $(VAR) since it is not of string type.
 			return in, nil
 		}
 		// This field can potentially contain a $(VAR) since it is

--- a/api/internal/accumulator/refvartransformer_test.go
+++ b/api/internal/accumulator/refvartransformer_test.go
@@ -29,6 +29,7 @@ func TestRefVarTransformer(t *testing.T) {
 		description string
 		given       given
 		expected    expected
+		errMessage  string
 	}{
 		{
 			description: "var replacement in map[string]",
@@ -111,6 +112,27 @@ func TestRefVarTransformer(t *testing.T) {
 				unused: []string{"BAR"},
 			},
 		},
+		{
+			description: "var replacement panic in map[string]",
+			given: given{
+				varMap: map[string]interface{}{},
+				fs: []types.FieldSpec{
+					{Gvk: resid.Gvk{Version: "v1", Kind: "ConfigMap"}, Path: "data/slice"},
+				},
+				res: resmaptest_test.NewRmBuilder(
+					t, resource.NewFactory(kunstruct.NewKunstructuredFactoryImpl())).
+					Add(map[string]interface{}{
+						"apiVersion": "v1",
+						"kind":       "ConfigMap",
+						"metadata": map[string]interface{}{
+							"name": "cm1",
+						},
+						"data": map[string]interface{}{
+							"slice": []interface{}{5}, // noticeably *not* a []string
+						}}).ResMap(),
+			},
+			errMessage: "expected array of strings, found [5]",
+		},
 	}
 
 	for _, tc := range testCases {
@@ -122,16 +144,23 @@ func TestRefVarTransformer(t *testing.T) {
 			err := tr.Transform(tc.given.res)
 
 			// assert
-			if err != nil {
-				t.Errorf("unexpected error: %v", err)
-			}
+			if tc.errMessage != "" {
+				if err == nil {
+					t.Fatalf("missing expected error %v", tc.errMessage)
+				} else if err.Error() != tc.errMessage {
+					t.Fatalf("actual error doesn't match expected error: \nACTUAL: %v\nEXPECTED: %v", err.Error(), tc.errMessage)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
 
-			a, e := tc.given.res, tc.expected.res
-			if !reflect.DeepEqual(a, e) {
-				err = e.ErrorIfNotEqualLists(a)
-				t.Fatalf("actual doesn't match expected: \nACTUAL:\n%v\nEXPECTED:\n%v\nERR: %v", a, e, err)
+				a, e := tc.given.res, tc.expected.res
+				if !reflect.DeepEqual(a, e) {
+					err = e.ErrorIfNotEqualLists(a)
+					t.Fatalf("actual doesn't match expected: \nACTUAL:\n%v\nEXPECTED:\n%v\nERR: %v", a, e, err)
+				}
 			}
-
 		})
 	}
 }

